### PR TITLE
TCP N11: query parameter binding — {name:Type} values over the native protocol

### PR DESCRIPTION
**Pull request by alex-clickhouse**
**Stacked on #560** (`tcp/epic-m-connection-pool`) — review that one first; this PR's diff is the parameters only.

TCP epic **N11**: bind values to a parameterized query's `{name:Type}` placeholders. The Query packet has carried a parameter-list field since L4, but the client always sent it empty.

```csharp
await using var client = new ClickHouseTcpClient("Host=localhost");

var options = new ClickHouseTcpQueryOptions
{
    Parameters = new ClickHouseTcpParameterCollection
    {
        { "since", new DateOnly(2026, 1, 1) },
        { "tags",  new[] { "a", "b" } },
    },
};

await foreach (object[] row in client.QueryAsync(
    "SELECT * FROM events WHERE day >= {since:Date} AND tag IN {tags:Array(String)}", options))
{
}
```

No method signature changed — every overload already took an options object, so only four internal `parameters: null` call sites moved. `InsertAsync` inherits `Parameters` from the same record, so `INSERT INTO t SELECT ... WHERE x > {t:Int32}` works with no extra plumbing.

## The wire contract is not what the spec said

This is the part worth reading. Our native-protocol spec note (`ignored-docs/tcp/native-protocol.md`, not in the repo) claimed that *String* parameters must arrive single-quoted. The truth is broader and sharper, and I established it by probing ~60 cases against ClickHouse 26.6.1 on both transports before writing any code.

**Every type must be quoted.** The parameter rides the settings list with `flags = 0x02 (Custom)`, so the server runs `Field::restoreFromDump` on the text *before* the `{name:Type}` substitution ever sees it. An unquoted `42` for an `Int32` parameter fails with `Couldn't restore Field from dump: 42`, exactly as a bare `Alice` does for a `String`.

**The value therefore crosses two unescape stages**, and stage 2 turns out to be the *same* escaped-text reader that HTTP's `param_<name>` feeds — I confirmed that by curling the HTTP endpoint and getting byte-identical results. So the rule is:

```
wire value = QuoteSingle(Escape(T))     where T is the exact text HTTP sends
```

To send the three characters `a\b` the wire carries `'a\\\\b'`. A literal newline must travel as `'a\\nb'`, because stage 2 terminates on a raw newline or tab.

Two server behaviours here are silent rather than diagnosed, and both corrupt data:

- **Doubling a quote does not escape it.** `'a''b'` restores as `a`, no error. Only backslash escaping works.
- **Trailing text after the closing quote is ignored.** `'ab' XYZ` restores as `ab`.

Neither is reachable through this API — the formatter owns the quoting — but they are why the escaping is not negotiable, and why `ParameterText.Escape` is applied twice rather than once. The spec note and the L4 TODO entry are corrected.

## Decisions worth a look

**The formatter is a deliberate duplicate, and that is the uncomfortable part.** Since the inner text must match HTTP byte for byte, the ideal is one implementation. The assembly graph forbids it: `ClickHouse.Driver` already references `ClickHouse.Driver.Tcp`, so a reference back is circular, and `HttpParameterFormatter` is written against the RowBinary `ClickHouseType` tree, which this assembly does not have. Sharing would mean moving `ClickHouseType`, `TypeConverter` and `TypeSettings` into `ClickHouse.Driver.Common` while keeping the net6.0 HTTP build green — an epic, not a PR. So `TcpParameterFormatter` mirrors `HttpParameterFormatter` arm for arm, and **nothing enforces that today**. A CI parity test (one shared case table through both formatters, asserting `tcp == http.Escape().QuoteSingle()`) is filed as TODO **N11a** rather than built here. It is also the precondition for ever collapsing the two.

**Five defects found by comparing against the other two implementations.** I first wrote this port to reproduce two known HTTP bugs on the grounds that parity was the invariant. That was the wrong call — parity with a bug is still a bug — so they are fixed, and I then went looking for more by diffing this formatter's behaviour against the HTTP test corpus and against clickhouse-go's native parameter tests.

They are fixed **on the TCP side only**. This PR touches nothing outside `ClickHouse.Driver.Tcp` and its tests: HTTP is a shipped transport, and changing its behaviour does not belong in a TCP epic PR. An earlier revision of this branch did fix them there too; that is reverted.

Two were inherited HTTP defects. **One of the two has since been fixed on `main` independently**, by #483/#484, so it is no longer a divergence:

- **A `byte[]` bound to a `String` parameter was sent as the literal text `System.Byte[]`.** Only `FixedString` read the bytes. Silent corruption, and reachable in practice since `readStringsAsByteArrays` makes that a plausible round-trip. HTTP now reads the bytes too, and also accepts a `ReadOnlyMemory<byte>`; this branch matches both.
- **`Interval<Unit>` threw.** TCP now sends the underlying `Int64` count, and the server accepts it, so this was purely a missing arm. HTTP could not be fixed the same way regardless: the RowBinary tree has no Interval type at all, so an HTTP query fails while *resolving the name*, before any formatter runs. Adding it is a feature with a read path and a PublicAPI change (there is already a `TODO: following interval implementation` at `Types/BinaryTypeDecoder.cs:136`). This one is still a divergence.

Three more were this port being *behind* HTTP, which the comparison is what surfaced:

- **`{p:Json}` threw** — the arm matched only the uppercase `JSON`, though the server takes either spelling and reports both as `JSON`.
- **`QBit(T, N)` had no arm.**
- **The six geo names had no arm** (`Point`, `Ring`, `LineString`, `Polygon`, `MultiLineString`, `MultiPolygon`). They now format as the shape they stand for, mirroring HTTP, where `PointType : TupleType` and the rest are `ArrayType`.

All eight of those types now have live round-trips.

**Type resolution has two sources:** a non-empty `ClickHouseType` override, then the query `{name:Type}` hint. The scanner ignores comments and string literals. If neither source supplies a type, binding throws with guidance to declare one. CLR inference remains for future client-side `?`/`@` placeholders and for matching `Variant` alternatives, but native `{name:Type}` binding does not call it.

**Deliberately out of scope:** ADO-style `@name` rewriting (this client is not an ADO provider; revisit if a TCP-backed `DbConnection` lands) and the `IParameterFormatter` / `IParameterTypeResolver` hooks (both live in the HTTP assembly and cannot be referenced from here).

**No changelog entry, and no diff outside the TCP projects.** No commit in this epic series has touched `CHANGELOG.md`, since the TCP client is unreleased and `[Experimental]`, and that holds here. Say the word if you want the whole epic summarised there when it lands.

**The divergence list is much shorter than it was, because `main` closed most of it.** An earlier revision of this description counted eight divergences, six of them HTTP defects this port fixed only on its own side. #483/#484 has since fixed the two that silently corrupted data (`byte[]` bound to `String`, and a `byte[]` that is not valid UTF-8) and added `TimeOnly` binding, and the geo and `QBit` arms turned out to exist on HTTP already. What I can still see after this round:

- **`Interval`.** TCP sends the `Int64` count; an HTTP query fails while resolving the name, because the RowBinary tree has no Interval type. Closing it is the feature described above, not a formatter change.
- **A kinded `DateTime`/`DateTimeOffset` against a hint that declares no timezone.** TCP refuses it, HTTP sends it. Deliberate, and argued below under N11d-http.
- **The text a byte payload travels as**, which is a representation difference rather than a behavioural one. HTTP's `EscapeBytes` emits `\xHH` for every byte outside printable ASCII; this formatter tries strict UTF-8 first, so `héllo` stays readable and only genuinely invalid UTF-8 falls back to `\xHH`. The server reads back the same bytes either way, which is the property that matters, but the two do not produce identical text and a naive `tcp == http.Escape().QuoteSingle()` assertion would fail on it.

That last point is the reason the parity check filed as **N11a** still needs building: it has to compare what the server stores, not the text. Nothing enforces any of this today.

**A silent timezone defect, now refused rather than sent.** A bare `{d:DateTime}` hint declares no timezone, so the server reads the wall-clock text in `session_timezone`. The client wrote a kinded `DateTime`/`DateTimeOffset` as UTC, so on a non-UTC session the instant moved — nine hours for `Asia/Tokyo` — with no error. The client now throws instead, naming the value's kind, what the server would have done, and both ways out:

> Parameter 'd' (type DateTime): A DateTimeOffset names an instant, but the type declares no timezone, so the instant cannot be sent without loss. The server reads the value in its session timezone, which moves the instant when that is not UTC, and reports no error. Declare the timezone in the type — DateTime('UTC') — or pass a DateTime with Kind=Unspecified to send a wall-clock time for the server to read in its own timezone.

`Kind=Unspecified` with a bare hint **stays legal**, and that limit is the point. Unspecified is a wall clock with no instant attached, which is exactly what a type with no timezone carries — nothing is lost and the round-trip is faithful. Only a value that *names an instant* is refused, because only then does the client hold information the wire cannot carry.

**Epoch was the first candidate and the evidence killed it.** clickhouse-go sends epoch seconds, which is immune to `session_timezone` — I confirmed that. But it cannot carry the range: the server reads an integer of **four digits or fewer as a year**, so epoch `0`–`9999` (1970-01-01 00:00:00 to 02:46:39 UTC) is rejected with `value 1000 cannot be parsed as Date`, as is any negative epoch for `DateTime`. An encoding with a hole in it is not an encoding. The other candidate — formatting into the session timezone — needs the client to guess server state at format time and go silently wrong when the guess is stale. Refusing beat guessing.

**HTTP keeps the old behaviour, deliberately.** The distinction is not timidity: on HTTP this combination is *correct today for everyone on a UTC server*, which is the default. Throwing there would break working code for the majority to protect the non-UTC minority, and it is a shipped client. TCP is unreleased and `[Experimental]`, so establishing "declare your timezone" costs nobody. Filed as **N11d-http** with three graded options (log a warning / opt-in switch / throw), and a note that the kinded half has no HTTP test at all — that should be written first, because it fails silently today.

## Security

The `Identifier` arm is the only one that emits its value unescaped, so I checked what it can reach. The server substitutes it as a single AST identifier and applies its own backtick quoting, so it cannot break out:

| value for `{c:Identifier}` | result |
|---|---|
| `a) UNION ALL (SELECT 999` | `UNKNOWN_IDENTIFIER`, treated as one name |
| ``a`,`b`` | escaped to ``a\`,\`b``, one name |
| `system.tables` | one name — it cannot even cross a database |

Ordinary values are inert too: `' OR 1=1 --` round-trips as data, covered by a test at both the formatter and the live-server level. An `Identifier` parameter is still a "trusted name" surface in the same sense it is on HTTP — it selects a name, so a caller must not let an untrusted value choose which column is read — but it cannot inject syntax.

## Testing

The full net9.0 TCP suite passes (2,669 tests), including the live-server parameter cases covering every supported parameter type. Coverage on the new files: formatter 99.3%, extractor 98.9%, collection 100%, `MapPairs` 100%, `ParameterText` 100%, inference 95.6%. The one uncovered branch worth naming is a defensive `if (sql[startIndex] != '{')` guard in the extractor that its only call site makes unreachable.

The escaping expectations are hand-derived from the two-stage rule and then confirmed against a real server, so the unit and integration layers check each other rather than restating one belief twice. The character-class cases are the ones I would keep: a carriage return and a NUL need no escape while a newline and a tab do, which is easy to get wrong in either direction.

**The corner cases came from reading the other two implementations' suites rather than from imagination.** I inventoried what `ClickHouse.Driver.Tests` binds as an HTTP parameter and what clickhouse-go binds as a native one, then took everything either had that this did not. The additions a live server can settle:

- `NaN`, `+Infinity`, `-Infinity` — .NET and ClickHouse spell these differently and the server happens to accept both, which only a round-trip shows;
- a `Bool` inside a composite — the server takes `1`/`0` for a scalar `Bool` but rejects them for `Array(Bool)`, so a formatter emitting digits passes the scalar case and fails this one (clickhouse-go hit exactly this, issue #1891);
- the `Date` and `Date32` range ends, a pre-epoch `DateTime64`, and scale 9;
- an `Enum` by number rather than by label; decimals of 30 and 50 digits, past what a CLR `decimal` holds;
- negative and past-24h `Time`/`Time64`; a surrogate pair; a nine-element `ValueTuple` (`TRest` nesting);
- four more `Map` shapes — empty, a key needing escapes, a non-string key, one holding an array.

Non-zero-lower-bound and zero-length array axes are unit tests, being shapes no server round-trip can produce.

**The suite can now gate on server version**, which QBit forced: it reached the matrix and failed on 25.8, which predates the type. `TcpFeature` carries a `[SinceVersion]` per capability and `TcpServerFeatures` maps a version onto the set — the same shape as `Feature` and `ClickHouseFeatureMap` in the HTTP suite, kept as a separate copy because the TCP projects do not reference `ClickHouse.Driver`, which is what keeps the TCP CI job building only the two TCP projects. Whole tests use `[RequiresServerFeature]`; a single case of a `TestCaseSource` guards its `yield`, since a case that is never yielded cannot carry an attribute. The version comes from `CLICKHOUSE_VERSION` rather than from the server, because a case source is enumerated during discovery, before the fixture has started the container. A moving tag resolves to "everything supported", so a real gap fails loudly instead of being skipped in silence. The mapping has its own tests — a wrong answer there does not fail a test, it skips one.

The comparison also found gaps in the **HTTP** suite that I did not fix here, filed as **N11c**: no NaN/Infinity parameter anywhere, no upper-bound dates (only the lower bound is ever tested), no DST-gap or DST-overlap instant, and `Time`/`Time64` commented out of the composite matrix. Each is two tests, and the HTTP half may well find something.

## Third review round

Five findings, plus a sixth that followed from fixing one of them. Four were this branch's to fix, one belongs to #609, and two carried over from earlier rounds are closed by argument rather than by code.

The largest was not really one finding. **The SQL scanner was a port of a pre-#509 version of the HTTP one**, so it re-introduced the bug #511 had already fixed there: the colon search reached past its own brace and took the colon of the next parameter, dropping that parameter's hint. `SELECT 1 AS "col{x}", {p:Int32}` is a query the server answers, and this client could not bind it. That was the reported symptom; the port was also missing backtick and double-quote identifiers, `$tag$` heredocs, `//` comments, nested block comments, and the rule that only `# ` and `#!` open a comment. The scanner is now the current HTTP one, ported whole, and each of those rules is pinned by a case confirmed against 26.6.1 first.

The comment rule cost one existing test, which asserted that a bare `#` opens a comment. The server disagrees: `SELECT {val:Int32} #{val:String}` fails with `Unrecognized token`, while `# ` and `#!` both work. That case is replaced by one pinning the server's rule, mirroring `ExtractTypeHints_BareHash_NotTreatedAsComment` in the HTTP suite. It is the only existing assertion this PR changes.

Three more defects:

- **A `ReadOnlyMemory<byte>` bound to `String` or `FixedString` was sent as the CLR type name.** HTTP reads it. Fixing the formatter exposed a second half, reported once the first fix was pushed: `Accepts` did not recognise a byte payload in one either, so no text `Variant` alternative would take it, and a `JSON` alternative would take it and serialize it as base64. Both halves are fixed.
- **A `byte[]` matched any `Array` alternative by name.** `Variant(Array(String), Array(UInt8))` sent `['65','66']`, which the server stores without complaint as a valid `Array(String)`. The element type decides now.
- **Geo names and `QBit` matched no `Variant` alternative**, though the formatter writes both, so `Variant(Point, String)` refused a value it could have written. They now match the shape the name stands for, which also made the geo shape mapping shared between formatting and matching instead of living only in the formatter.

**`TimeOnly` binding belongs to #609**, which lands the same three hunks. I had written a duplicate and removed it before committing, so this branch carries none and the stack has exactly one.

Two findings are declined:

- **`FormatTime64` can emit `0:00:60`.** Checked against 26.6.1: `CAST('0:00:60' AS Time64(0))` normalizes to `00:01:00`, and the HTTP formatter emits the same text. Non-canonical, not rejected and not corrupting.
- **`Variant` matching enumerates a sequence twice**, once to choose an alternative and once to format. Buffering would add an allocation and a full traversal to the normal path. The contract is now written on `ClickHouseTcpParameter.Value` rather than living only in a review reply.

Every new test here was run against the unfixed code first: 21 unit and 7 integration cases fail without the scanner, byte and geo fixes, and 4 more fail without the `ReadOnlyMemory` matcher.

## Second review round

Six inline comments, all valid against current `HEAD` — none of them outdated. Four defects, plus two of my own tests that passed against the code they were meant to guard:

- **The SQL scanner ignored backslash-escaped quotes.** A quoted literal could appear to end early, hiding a later placeholder and losing its type hint.
- **Variant alternatives were matched on the outer type name alone,** so a `string[]` took the first arm of `Variant(Array(Int32), Array(String))` and every element went out unquoted. Matching now recurses into element, key/value and tuple element types, and checks tuple arity.
- **A Map read back from this client could not be bound again.** The codec surfaces a row as `KeyValuePair[]` so duplicate keys and order survive, and only `IDictionary` was accepted.
- **`byte[]` was decoded as UTF-8,** so `0xFF` became U+FFFD and the server stored `EF BF BD`. Probing showed `\xHH` in the inner text yields the exact byte, so bytes that are not valid UTF-8 now travel that way while valid UTF-8 keeps the readable form.
- **The `InsertAsync` test called `ExecuteAsync`.** A dropped `parameters` argument in either real overload would have passed the whole suite. Both are covered now, through a parameterized target.
- **The one-shot sequence test exercised a fallback that native binding no longer uses.** Native parameters now require an explicit type or SQL hint. Inference remains directly tested for future client-side placeholders.

Since the sharpest finding was "this test passes on the unfixed code", I reverted all four source fixes and re-ran. Fifteen new tests failed, as they should — and **two of mine passed**, so I had repeated the mistake while fixing it. One put the literal after the placeholder, so the scanner never entered a string on the way to the hint; with that corrected it still passed, because an even number of escaped quotes makes the mis-scan open and close the same number of times and land outside a string again. The third attempt showed the difference is not observable through the server at all, so that test is gone and a comment points at the unit tests, which are true negatives. Every new test here has been run against the unfixed code and fails there.

Coverage then flagged that the Variant pair-sequence arm had no test at all, and that a `try/catch (FormatException)` I had added around `NamedElementParser.Split` can never fire — the exception comes from `ElementTypeStrings`, which validates arity. Both fixed; `ParameterTypeInference` 82% → 95%.

## Review outcomes

This is the earlier review pass, before the cross-implementation comparison above. Five defects found and fixed. Four were parity divergences, and one of those was silent:

- **A `byte[]` in a `Variant` became the text `System.Byte[]`.** It matched a string alternative before an array one, and the string arm prints the CLR type. HTTP picks `Array(UInt8)` via `CanWrite`. This was the only silent corruption found — everything else failed loudly.
- **`Time64` rounded away from zero** where HTTP pre-rounds to even, so a midpoint landed one tick high.
- **A decimal given as text** rejected the exponent, thousands-separator and accounting-parentheses forms HTTP accepts.
- **A `Variant` error named a parameter the caller never wrote**, because alternative matching inferred under a placeholder name.
- **The native path inferred untyped parameters and could consume one-shot sequences.** Native binding now requires a declared type, so it never takes that path; inference remains for future client-side placeholders.

Plus the multidimensional-array depth check I found myself before the review landed: the arm peeled the declared nesting only as far as the CLR rank, so a rank-3 array declared `Array(Array(T))` emitted three bracket levels and only the server noticed.

**One trap the client cannot fix, and CI corrected my understanding of it.** A parameter named after a server setting can be applied *as that setting*, because the native protocol carries parameters in the settings list. `limit` and `offset` are the names callers reach for, and the failure is a parse error about a quoted string that names neither the parameter nor the cause.

I first wrote this up as a fixed protocol limitation, on the evidence of my local 26.6.1. The matrix disagreed: **it is version-dependent and already fixed upstream** — 25.8 through 26.6 reject the query, `latest` binds correctly. My local server happened to sit on the broken side, so nothing short of the matrix would have shown it.

The test therefore pins the property that holds on *every* version rather than one version's error: such a name must **never bind to the wrong value**. Either the server rejects the query, or the count is right; a wrong count fails. That is the outcome that would actually hurt, and it is a stronger assertion than the one it replaced. The doc now says "rename the parameter if you support any server in 25.8–26.6" instead of claiming the protocol forbids it. Unaffected on HTTP, which carries the name separately.

The review also confirmed the parts I most wanted checked: the two-stage escaping held against every adversarial input it tried (BEL/BS/FF/VT, an already-quoted value, a 200 000-character value, literal `\x41` and `\N` sequences, emoji), and the `Identifier` arm is contained by the outer escape. It also called the extractor a faithful port, which the third round showed it was not: faithful to the HTTP scanner as it stood before #509, three fixes behind.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: ClickHouse.Driver.Tcp/Parameters/TcpParameterFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+            // Keep pair sequences out of the general Array arm while preferring dictionaries.
+            case "Map" when type.Arguments.Count == 2 && MapPairs.IsPairSequence(value):
+                return FormatMapPairs(type, (IEnumerable)value);
+
+            case "Variant":
+                return FormatVariant(type, value, quote);
+
+            // The server takes either spelling and reports the type as JSON, so both must format.
+            case "JSON" or "Json":
+                return value is string json ? json : JsonSerializer.Serialize(value);
```
[the remark this answers is not in the document]
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp/Parameters/TcpParameterFormatter.cs]: Fixed in 4054b2fb. JSON text now gets the inner escape layer before the outer Field escaping. A live round-trip with raw newlines and tabs verifies that the complete document reaches the server.

File: ClickHouse.Driver.Tcp/Parameters/TcpParameterFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+        int hours = (int)(absolute / 3600m);
+        decimal remainder = absolute % 3600m;
+        int minutes = (int)(remainder / 60m);
+        decimal seconds = remainder % 60m;
+
+        // Pre-round to even because decimal formatting rounds midpoint digits away from zero.
+        string secondsText = Math.Round(seconds, scale, MidpointRounding.ToEven)
+            .ToString("00." + new string('0', scale), CultureInfo.InvariantCulture);
+        string text = $"{hours}:{minutes:D2}:{secondsText}";
+        return negative ? "-" + text : text;
```
[the remark this answers is not in the document]
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp/Parameters/TcpParameterFormatter.cs:576]: No change here. I verified this against ClickHouse 26.6.1: CAST of 0:00:60 as Time64(0) is accepted and normalizes to 00:01:00. The shipped HTTP formatter emits the same form, so this is non-canonical text rather than rejection or corruption.

File: ClickHouse.Driver.Tcp/Parameters/ParameterTypeInference.cs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+            // These share one CLR type with several ClickHouse types, so the base name alone decides.
+            string or char => node.Name is "String" or "FixedString" or "Enum8" or "Enum16" ? node.Name : "String",
+            DateTime or DateTimeOffset => node.Name is "DateTime" or "DateTime64" or "Date" or "Date32" ? node.Name : "DateTime64",
+            decimal or ClickHouseDecimal => node.Name.StartsWith("Decimal", StringComparison.Ordinal) ? node.Name : "Decimal128",
+            not string and IEnumerable => "Array",
+            _ => InferOrNothing(value),
+        };
+
+        return string.Equals(inferred, node.Name, StringComparison.Ordinal);
```
[the remark this answers is not in the document]
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp/Parameters/ParameterTypeInference.cs:147]: Fixed in 4054b2fb. JSON is now deferred as the broad fallback after concrete alternatives. Tests cover JSON fallback, a live Variant(JSON, UInt64) round-trip, and Map winning when JSON is declared first.

File: ClickHouse.Driver.Tcp/Parameters/TcpParameterFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+        {
+            return quote ? "null" : NullValueString;
+        }
+
+        foreach (TypeNode alternative in type.Arguments)
+        {
+            if (ParameterTypeInference.Accepts(alternative, value))
+            {
+                return Format(alternative, value, quote);
+            }
```
[the remark this answers is not in the document]
Comment (alex-clickhouse) [reply ClickHouse.Driver.Tcp/Parameters/TcpParameterFormatter.cs:539]: Leaving repeated enumeration as the explicit contract. Variant matching may inspect a sequence before formatting it; callers should provide a replayable collection or array. Buffering here added allocation and a full traversal to the normal path, and it could change JSON fallback semantics.

File: ClickHouse.Driver.Tcp.Tests/Integration/ClickHouseTcpParameterIntegrationTests.cs [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+// Live-server coverage for the two parsing stages native parameter values cross.
+[TestFixture]
+[Category("Integration")]
+public class ClickHouseTcpParameterIntegrationTests
+{
+    private static readonly CancellationToken None = CancellationToken.None;
+
+    private static IEnumerable<TestCaseData> RoundTripCases()
+    {
+        yield return new TestCaseData("Int32", 42).Returns("42").SetName("Int32");
```
Comment (kavirajk) [inline ClickHouse.Driver.Tcp.Tests/Integration/ClickHouseTcpParameterIntegrationTests.cs:25]: Looks like Bfloat16 test cases are missing.

**Review by kavirajk — APPROVED** [review]
LGTM 👍
